artifactory/utils/container/buildinfo.go

```diff
@@ -517,73 +277,154 @@ func removeDuplicateLayers(imageMLayers []layer) []layer {
-		return 0, nil
+func toNoneMarkerLayer(layer string) string {
+	imageId := strings.Replace(layer, "__", ":", 1)
+	return strings.Replace(imageId, ".marker", "", 1)
+}
+
+type CommandType string
+
+// Create a image's build info from manifest.json.
+func (builder *buildInfoBuilder) createBuildInfo(commandType CommandType, manifest *manifest, candidateLayers map[string]*utils.ResultItem, module string) (*buildinfo.BuildInfo, error) {
+	imageProperties := map[string]string{
+		"docker.image.id":  builder.imageSha2,
+		"docker.image.tag": builder.image.Name(),
 	}
-	totalDownloaded := 0
-	remoteRepo := builder.repositoryDetails.key
-	imageName := getImageName(builder.image.Tag())
-	clientDetails := builder.serviceManager.GetConfig().GetServiceDetails().CreateHttpClientDetails()
-	// Search for marker layers
-	for _, layerData := range resultMap {
-		if strings.HasSuffix(layerData.Name, markerLayerSuffix) {
-			log.Debug(fmt.Sprintf("Downloading %s layer into remote repository cache...", layerData.Name))
-			baseUrl := builder.serviceManager.GetConfig().GetServiceDetails().GetUrl()
-			endpoint := "api/docker/" + remoteRepo + "/v2/" + imageName + "/blobs/" + toNoneMarkerLayer(layerData.Name)
-			resp, body, err := builder.serviceManager.Client().SendHead(baseUrl+endpoint, &clientDetails)
-			if err != nil {
-				return totalDownloaded, err
-			}
-			if resp.StatusCode != http.StatusOK {
-				return totalDownloaded, errorutils.CheckErrorf("Artifactory response: " + resp.Status + "for" + string(body))
+	if module == "" {
+		imageName, err := builder.image.GetImageBaseNameWithTag()
+		if err != nil {
+			return nil, err
+		}
+		module = imageName
+	}
+	// Manifest may hold 'empty layers'. As a result, promotion will fail to promote the same layer more than once.
+	manifest.Layers = removeDuplicateLayers(manifest.Layers)
+	var artifacts []buildinfo.Artifact
+	var dependencies []buildinfo.Dependency
+	var err error
+	switch commandType {
+	case Pull:
+		dependencies, err = builder.createPullBuildProperties(manifest, candidateLayers)
+	case Push:
+		artifacts, dependencies, builder.imageLayers, err = builder.createPushBuildProperties(manifest, candidateLayers)
+		if err != nil {
+			return nil, err
+		}
+		if !builder.dryRun {
+			if err := setBuildProperties(builder.buildName, builder.buildNumber, builder.project, builder.imageLayers, builder.serviceManager); err != nil {
+				return nil, err
 			}
-			totalDownloaded++
 		}
 	}
-	return totalDownloaded, nil
+	buildInfo := &buildinfo.BuildInfo{Modules: []buildinfo.Module{{
+		Id:           module,
+		Type:         buildinfo.Docker,
+		Properties:   imageProperties,
+		Artifacts:    artifacts,
+		Dependencies: dependencies,
+	}}}
+	return buildInfo, nil
 }
 
-func getImageName(image string) string {
-	imageId, tag := strings.LastIndex(image, "/"), strings.LastIndex(image, ":")
-	if imageId == -1 || tag == -1 {
-		return ""
+// Create a image's build info from list.manifest.json.
```

Comment:
Create a image's build info from list.manifest.json.
-->
Create the image's build info from list.manifest.json.